### ISSUE TITLE
Stack unwinding is broken in Linux 6.4 on x86-64 due to ORC format change

### ISSUE THREAD

**Issue by osandov** [osandov/drgn#303]
```
$ python3 -m vmtest.vm -k '6.4.*' python3 -Bm drgn
...
>>> prog.stack_trace(1)
Traceback (most recent call last):
  File "/usr/lib64/python3.11/code.py", line 90, in runcode
    exec(code, self.locals)
  File "<console>", line 1, in <module>
Exception: unknown ORC entry type 3
```

This is because torvalds/linux@fb799447ae2974a07907906dff5bd4b9e47b7123 changed the ORC entry format: the type is now 3 bits, and types now range from 0-4. torvalds/linux@ffb1b4a41016295e298409c9dbcacd55680bd6d4 also previously changed the ORC format in a way that we didn't detect.

We might be able to guess the "version" of the format, but let's see if we can make upstream include a version identifier first.

**Comment by brenns10** [comment]
According to the [docs](https://drgn.readthedocs.io/en/latest/advanced_usage.html#environment-variables), ORC should not be used unless either (a) `DRGN_PREFER_ORC_UNWINDER=1`, or (b) there is no DWARF call frame information. I'm curious why the vmtest encountered this issue? The vmtest kernels have DWARF info, so do they run with `DRGN_PREFER_ORC_UNWINDER=1` set?

**Comment by osandov** [comment]
drgn falls back to using ORC if it can't find DWARF CFI for a given program counter. Most kernel stack traces end up with a frame without DWARF CFI because the kernel entry code is in assembly. For example, in this stack trace:

```
>>> prog.stack_trace(1)
#0  context_switch (./kernel/sched/core.c:5299:2)
#1  __schedule (./kernel/sched/core.c:6612:8)
#2  schedule (./kernel/sched/core.c:6688:3)
#3  do_wait (./kernel/exit.c:1631:4)
#4  kernel_wait4 (./kernel/exit.c:1775:8)
#5  __do_sys_wait4 (./kernel/exit.c:1803:13)
#6  do_syscall_x64 (./arch/x86/entry/common.c:50:14)
#7  do_syscall_64 (./arch/x86/entry/common.c:80:7)
#8  entry_SYSCALL_64+0xae/0x1aa (./arch/x86/entry/entry_64.S:120)
#9  0x7f00391eac37
```

`entry_SYSCALL_64` doesn't have DWARF, so we used ORC and recovered the saved userspace registers as frame #9.

**Comment by brenns10** [comment]
Ah interesting, I hadn't gotten the idea that there's sometimes not DWARF for the function at the bottom of the stack. Maybe I should try out dwarfdump to see what data is present for what functions, it looks capable of doing that.

**Comment by osandov** [comment]
Closed by 91ede0c6a440bc2178aa2d8776d4a0caef952121. Hopefully https://lore.kernel.org/linux-debuggers/aef9c8dc43915b886a8c48509a12ec1b006ca1ca.1686690801.git.osandov@osandov.com/ is merged so that the terrible kernel version check is limited.

**Comment by osandov** [comment]
My patch was merged in torvalds/linux@b9f174c811e3ae4ae8959dc57e6adb9990e913f4.
